python/ucxx/_lib_async/continuous_ucx_progress.py

```diff
@@ -70,3 +72,50 @@ async def _progress_task(self):
             worker.progress()
             # Give other co-routines a chance to run.
             await asyncio.sleep(0)
+
+
+class BlockingMode(ProgressTask):
+    def __init__(self, worker, event_loop, epoll_fd):
+        super().__init__(worker, event_loop)
+
+        # Creating a job that is ready straightaway but with low priority.
+        # Calling `await self.event_loop.sock_recv(self.rsock, 1)` will
+        # return when all non-IO tasks are finished.
+        # See <https://stackoverflow.com/a/48491563>.
+        self.rsock, wsock = socket.socketpair()
+        self.rsock.setblocking(0)
+        wsock.setblocking(0)
+        wsock.close()
+
+        # Bind an asyncio reader to a UCX epoll file descripter
+        event_loop.add_reader(epoll_fd, self._fd_reader_callback)
+
+        # Remove the reader and close socket on finalization
+        weakref.finalize(self, event_loop.remove_reader, epoll_fd)
+        weakref.finalize(self, self.rsock.close)
+
+    def _fd_reader_callback(self):
+        self.worker.progress()
+
+        # Notice, we can safely overwrite `self.dangling_arm_task`
+        # since previous arm task is finished by now.
+        assert self.asyncio_task is None or self.asyncio_task.done()
+        self.asyncio_task = self.event_loop.create_task(self._arm_worker())
+
+    async def _arm_worker(self):
+        # When arming the worker, the following must be true:
+        #  - No more progress in UCX (see doc of ucp_worker_arm())
+        #  - All asyncio tasks that isn't waiting on UCX must be executed
+        #    so that the asyncio's next state is epoll wait.
+        #    See <https://github.com/rapidsai/ucx-py/issues/413>
```

Comment:
Although I think I understand the constraint, I am not sure what it means for the "next state" to be epoll_wait. Surely there can be arbitrary non-ucx tasks?

Reply:
I'm gonna be honest and say my understanding here is also a bit fuzzy, and as you noted yourself this "doesn't work" (except when it does), with the original being adapted from https://stackoverflow.com/a/48491563. In my understanding, what `epoll_wait` refers to here is the [socket state](https://github.com/rapidsai/ucxx/pull/116/files#diff-1b56e195b331ac6ca2c996f330a9392fddaf6f124da76d3c290b97cabd57d2daR85-R88), which is there solely to provide a mechanism to prevent `asyncio` from running out of "ready" tasks, so `epoll_wait` will ensure that if nothing useful happens in the event loop, `asyncio` will still be awaken at some point to allow the loop to reevaluate.